A4A > Feedback: Implement Churn Feedback Mechanism

yashwin opened the pull request:
Resolves https://github.com/Automattic/automattic-for-agencies-dev/issues/2013

## Proposed Changes

* Implement Churn Feedback Mechanism while revoking licenses (Product & Hosting)

Note: The Slack channel name can be found here: https://github.com/Automattic/automattic-for-agencies-dev/issues/2006).

## Why are these changes being made?

* To help bolster our retention and get relevant, in-context feedback from agencies.

## Testing Instructions

* Patch 180225-ghe-Automattic/wpcom to your sandbox
* Open the A4A live link
* Go to Purchases > Licenses: Revoke a license for all the mentioned scenarios below and verify it works as expected.

----

Revoke a product not assigned to a site > Verify the below popup is shown > Fill out the form > Click the `Cancel license` button > Verify the license is revoked and the Slack message is published for the survey.

<img width="811" alt="Screenshot 2025-04-09 at 3 18 04 PM" src="https://github.com/user-attachments/assets/caaf1c81-da0b-4c05-ab5d-55cd4389e614" />

---

Revoke a product assigned to a site > Verify the below popup is shown > Fill out the form >  Click the `Cancel License` button > Verify you are redirected to WP.com and the Slack message is published for the survey. Click

<img width="811" alt="Screenshot 2025-04-09 at 3 21 37 PM" src="https://github.com/user-attachments/assets/30ab82ba-119b-42c0-8769-86e15181b9d5" />

---

Revoke a WP.com hosting license not assigned to a site > Verify the below popup is shown > Don't fill out the form > Click the `Cancel license` button > Verify the license is revoked and the Slack message is **not** published for the survey.

<img width="811" alt="Screenshot 2025-04-09 at 3 25 03 PM" src="https://github.com/user-attachments/assets/37222337-aa76-4282-9c2e-9e851e5a29a3" />

---

Revoke a WP.com hosting license assigned to a site > Verify the below popup is shown > Don't fill out the form > Click the `Speak with my Partner Manager` button > Verify the license is **not** revoked and the Slack message is published for the survey.

<img width="811" alt="Screenshot 2025-04-09 at 3 25 43 PM" src="https://github.com/user-attachments/assets/bd7ff6b9-9195-4df9-80a2-a4664582738c" />

---

Revoke a Pressable hosting license > Verify the below popup is shown > Fill out the form > Click the `Speak with my Partner Manager` button > Verify the license is **not** revoked and the Slack message is published for the survey.

<img width="811" alt="Screenshot 2025-04-09 at 3 28 58 PM" src="https://github.com/user-attachments/assets/f7206091-2a11-4730-9767-f27c5286e13a" />

---

Revoke a bundle product license > Verify the below popup is shown > Fill out the form > Click the `Cancel license` button > Verify the license is revoked and the Slack message is published for the survey.

<img width="811" alt="Screenshot 2025-04-09 at 3 30 24 PM" src="https://github.com/user-attachments/assets/1e9bba84-e2d8-4142-90b2-1e5363baad70" />

<img width="523" alt="Screenshot 2025-04-15 at 11 19 18 AM" src="https://github.com/user-attachments/assets/c4dd1034-bc19-4383-8ef6-47e5f4dea1cd" />


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
implement contextal feedback for post-purchase flow

yashwin committed:
minor ui fix

yashwin committed:
create feedback section and update logic

yashwin committed:
A4A: Feedback: Implement Churn Mechanism

yashwin committed:
Add custom icon support for SimpleList component

yashwin committed:
Use the CancelLicenseFeedbackModal when revoking a license

yashwin committed:
minor ui fix

yashwin committed:
use the url slug for site

jkguidaven commented:
Could we include padding in the inner container to prevent the cut-off seen below?

<img width="872" alt="Screenshot 2025-04-09 at 6 59 17 PM" src="https://github.com/user-attachments/assets/3a262557-15b3-4204-a206-e19afa2cdd63" />

jkguidaven commented on client/a8c-for-agencies/components/a4a-feedback/churn-mechanism/cancel-license-feedback-modal.tsx:
```diff
@@ -0,0 +1,333 @@
+
+	useEffect( () => {
+		// If the user clicks cancel and the license is revoked successfully, show a success notice
+		if ( selectedCTA === 'cancel' && revokeLicenseStatus === 'success' ) {
+			dispatch(
+				successNotice( translate( 'License revoked successfully' ), {
+					displayOnNextPage: true,
+					id: 'revoke-license-success',
+				} )
+			);
+			// Refresh the license list and close the modal if the user has not provided any feedback
+			if ( ! isFeedbackValid( 'cancel' ) ) {
+				handleClose();
+			}
+		}
+		// If the user clicks speak with partner manager and the feedback is submitted successfully,
+		// show a success notice and close the modal
+		if ( selectedCTA === 'speak-with-partner-manager' && saveFeedbackStatus === 'success' ) {
+			dispatch(
+				successNotice( translate( 'Thanks! We will connect you with a partner manager.' ), {
+					displayOnNextPage: true,
+					id: 'submit-product-feedback-success',
+				} )
+			);
+			onClose();
+		}
```
I'm not sure about the necessity of `useEffect` in this situation. Can't we simply move this logic to the click handler? Keeping all the logic in one location is more readable, in my opinion, than having it separated.

yashwin replied:
The thing is that we are triggering both API calls together. We don't want to close the modal before both are done. Also, not both APIs are called always, which makes us take this route.

jkguidaven commented on client/a8c-for-agencies/components/a4a-feedback/churn-mechanism/cancel-license-feedback-modal.tsx:
```diff
@@ -0,0 +1,333 @@
+				} )
+			);
+
+			saveFeedback( { params } );
```
Regarding my comment about the `useEffect` below, can we place the post-success logic here to make it easier to follow?

```

			saveFeedback(
				{ params },
				{
					onSuccess: () => {
						// DO success here.
					},
				}
			);
```

yashwin replied:
Replied in the above conversation.

jkguidaven commented on client/a8c-for-agencies/components/a4a-feedback/churn-mechanism/cancel-license-feedback-modal.tsx:
```diff
@@ -0,0 +1,333 @@
+								<CheckboxControl
+									key={ `suggestion-${ option.value }` }
+									label={ option.label }
+									checked={
+										!! suggestions.find( ( suggestion ) => suggestion.value === option.value )
+									}
```
```suggestion
									checked={ suggestions.some(
										( suggestion ) => suggestion.value === option.value
									) }
```

We can utilize `array.some` here so we don't need to convert the return type to boolean. It's also more natural to read.

yashwin replied:
Sure! Updated: https://github.com/Automattic/wp-calypso/pull/102496/commits/8d9d435a450b3bc05da8a2b69bf15a70b2f6311f

jkguidaven reviewed (commented):
Overall it works great. I left bunch of comments. Let me know what you think.

jkguidaven commented on client/a8c-for-agencies/components/a4a-feedback/churn-mechanism/cancel-license-feedback-modal.tsx:
```diff
@@ -0,0 +1,333 @@
+				survey_responses: {
+					comment: { text: comments },
+					suggestions: {
+						text: suggestions?.map( ( suggestion ) => suggestion.value ).join( ', ' ) || '',
```
**Nitpicking**: We should prefer the `??` nullish coalescing operator over `||`, as it is safer. In this instance, however, `||` works perfectly fine. Hence my comment is nitpick.

yashwin replied:
Updated it here: https://github.com/Automattic/wp-calypso/pull/102496/commits/8d9d435a450b3bc05da8a2b69bf15a70b2f6311f

jkguidaven commented on client/a8c-for-agencies/sections/purchases/licenses/license-preview/license-bundle-dropdown.tsx:
```diff
@@ -2,22 +2,27 @@ import { Gridicon, Button } from '@automattic/components';
-export default function LicenseBundleDropDown( { licenseKey, product, bundleSize }: Props ) {
+export default function LicenseBundleDropDown( {
+	licenseKey,
+	product,
```
While we're at it, can we rename this prop to productName instead? This is a bit confusing as it seems like a product object, but actually, it's just the product name.

yashwin replied:
Updated it here: https://github.com/Automattic/wp-calypso/pull/102496/commits/8d9d435a450b3bc05da8a2b69bf15a70b2f6311f

yashwin committed:
address PR comment

yashwin committed:
send product meta

yashwin commented:
Thanks for the review, @jkguidaven & @madebynoam!

@jkguidaven:

I have addressed your comments and made the changes accordingly: https://github.com/Automattic/wp-calypso/pull/102496/commits/8d9d435a450b3bc05da8a2b69bf15a70b2f6311f

I have also made some changes to the support product info as per your comment here - 180074-ghe-Automattic/wpcom: https://github.com/Automattic/wp-calypso/pull/102496/commits/7453d5ad298856e82885ef2f09a84c2126585183

jkguidaven reviewed (approved):
Thanks for addressing my comments. LGTM 👍 

Niced work. 👏

jeffgolenski commented:
Nice @yashwin! These are working nicely. Just a couple of things in terms of UX:

1. When a product, bundle, or hosting is canceled, in the slack message, can you display the name of the thing being canceled? Might be prudent to just replace "rating" in this instance. 
![CleanShot 2025-04-14 at 12 37 05](https://github.com/user-attachments/assets/37c6b185-13e0-48e6-9fed-83a5b3ed9c19)

2. Can the "and more" links all open a new tab so the user isn't taken away from this modal?
![CleanShot 2025-04-14 at 12 37 55](https://github.com/user-attachments/assets/acfca0b0-781f-4e23-810f-3463047f80ab)


3. Just for personalization sake can we say "your partner manager" instead of "a partner manager"?
![CleanShot 2025-04-14 at 12 38 54](https://github.com/user-attachments/assets/235d31b8-526a-476d-93e9-be198fde9f5a)

4. Biggest ask: "Speak with my partner manager" buttons - Let me know what is easily doable and what will need to be a second iteration. I want to ensure we're setting expectations correctly. 

![CleanShot 2025-04-14 at 12 39 17](https://github.com/user-attachments/assets/026f66b8-c102-4229-96c9-f3a868824855)


**Ideal scenario:** User clicks this and the request is sent to slack and directly pings the corresponding PM of that agency in their channel. IE: Agency X clicks "Speak with my partner manager" and Ross is pinged to contact them in #a4a-pm-ross

**MVP scenario:** We change to "Speak with my partner manager" to "Contact us" and submit a request to zendesk with a high priority.

jkguidaven commented:
> Nice @yashwin! These are working nicely. Just a couple of things in terms of UX:
> 
> 1. When a product, bundle, or hosting is canceled, in the slack message, can you display the name of the thing being canceled? Might be prudent to just replace "rating" in this instance.
>    ![CleanShot 2025-04-14 at 12 37 05](https://private-user-images.githubusercontent.com/214813/433468962-37c6b185-13e0-48e6-9fed-83a5b3ed9c19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NTAzOTIsIm5iZiI6MTc0NDY1MDA5MiwicGF0aCI6Ii8yMTQ4MTMvNDMzNDY4OTYyLTM3YzZiMTg1LTEzZTAtNDhlNi05ZmVkLTgzYTViM2VkOWMxOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQxNzAxMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOGYwOTRlYjk1ZDEwODIxZTYwYzIwOTRkZDVkNjMwMGQ3ZDc2ODFhYjliZTNmODk3N2YyMGY1MzRiMWVjMTU2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DIPi9rK33Tyn3MhMvvkpryuzJ1yJJhoPFjyNDJKztgU)

@jeffgolenski If I'm not mistaken, you haven't applied the backend patch yet. I believe here's how it would appear with the backend patch.

<img width="567" alt="Screenshot 2025-04-15 at 1 06 04 AM" src="https://github.com/user-attachments/assets/804f66d7-e6f2-43f1-9a34-6a20fa1f5041" />

yashwin committed:
address PR comments

yashwin commented:
> Nice @yashwin! These are working nicely. Just a couple of things in terms of UX:

Thanks for the review, @jeffgolenski!

> 1. When a product, bundle, or hosting is canceled, in the slack message, can you display the name of the thing being canceled? Might be prudent to just replace "rating" in this instance.

> We are already displaying this. You can check @jkguidaven's message: https://github.com/Automattic/wp-calypso/pull/102496#issuecomment-2802356120

> 2. Can the "and more" links all open a new tab so the user isn't taken away from this modal?

Done!

> 3. Just for personalization sake can we say "your partner manager" instead of "a partner manager"?

Good catch. Done! 

> 4. Biggest ask: "Speak with my partner manager" buttons - Let me know what is easily doable and what will need to be a second iteration. I want to ensure we're setting expectations correctly.

> **Ideal scenario:** User clicks this and the request is sent to slack and directly pings the corresponding PM of that agency in their channel. IE: Agency X clicks "Speak with my partner manager" and Ross is pinged to contact them in #a4a-pm-ross

We are already publishing these messages on the PM's channel. Basically, all the surveys are also published on the PM channel.

jeffgolenski commented:
> We are already publishing these messages on the PM's channel. Basically, all the surveys are also published on the PM channel

Ah got it. On second thought, I didn't see it come through those channels because my account isn't assigned a PM. Sorry about that!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17406162

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.
